mlir/lib/Catalyst/Transforms/SplitMultipleTapes.cpp

```diff
@@ -0,0 +1,332 @@
+        // preprocessing
+        // quantum.device[...]
+        // tape 1
+        // quantum.device_release
+        // quantum.device[...]
+        // tape 2
+        // quantum.device_release
+        // ... more tapes
+        // post processing and return
+        // }
+
+        // This function parses the operations in the funcop and puts them into
+        // the container as
+        // OpsEachTape = [[preprocessing ops], [tape1 ops], [tape2 ops], ..., [postprocessing ops]]
+
+        // special: [0] for pre and [-1] for post processing
+        unsigned int curTape = 0;
+
+        func->walk([&](Operation *op) {
+            if (op == func) {
+                return; // don't visit the funcop itself
+            }
+            if (isa<catalyst::quantum::DeviceInitOp>(op)) {
+                curTape++;
+            }
+            OpsEachTape[curTape].push_back(op);
+            if ((isa<catalyst::quantum::DeviceReleaseOp>(op)) &&
+                (curTape == OpsEachTape.size() - 2)) {
+                // reached post processing
+                curTape++;
+            }
+        });
+    } // collectOperationsForEachTape()
+
+    void collectNecessaryValuesFromEarlierTapes(
+        const std::vector<Operation *> &TapeOps,
+        SmallSet<std::pair<Operation *, unsigned int>> &NecessaryValuesFromEarlierTapes)
+    {
+        // Go through a list of operations and collect all the necessary operand values
+        // not defined in this list itself, and add them to NecessaryValuesFromEarlierTapes
+        // Record the Value as pair<defining op, result index number>
+        auto findNecessaryValues = [&](Operation *op) {
+            for (auto operand : op->getOperands()) {
+                if (isa<BlockArgument>(operand)) {
+                    // Arguments to the original multitape function are always available
+                    // and do not have defining ops, so don't process them
+                    continue;
+                }
+                Operation *OperandSource = operand.getDefiningOp();
+                if (std::find(TapeOps.begin(), TapeOps.end(), OperandSource) != TapeOps.end()) {
+                    continue;
+                }
+
+                // A list operand not defined in list itself, must be from earlier
+                // tapes/preprocessing!
+                unsigned whichResult = cast<mlir::OpResult>(operand).getResultNumber();
+
+                // Note that we should not record the same value twice in the set
+                // Enforce the pair as rvalue for `contains`
+                auto &&pair = std::make_pair(OperandSource, whichResult);
+                if (!NecessaryValuesFromEarlierTapes.contains(pair)) {
+                    NecessaryValuesFromEarlierTapes.insert(
+                        std::make_pair(OperandSource, whichResult));
+                }
+            }
+        };
+        for (Operation *PPOp : TapeOps) {
+            PPOp->walk(findNecessaryValues);
+        }
```

Comment:
I wonder if we can avoid the nested walks here. But this would also still be ok. I am just a bit worried about nested values which will never be used because we operate at a very high level, when all operations return values.

Reply:
The walker is to detect uses that are not exposed at the top-level operations in postprocessing operations. See the added test `Test when a post processing op is nested and the necessary values are not exposed at the upmost layer`